File: sort.rb
#! /usr/bin/ruby

def pr(list)
  printf("======================\n")
  list.each do |no|
     printf("%d\n", no)
  end
end

def mid3(x,y,z)
  if (x < y) then
    if (y < z) then return(y) elsif (z < x) then return(x) else return(z) end
  else 
    if (z < y) then return(y) elsif (x < z) then return(x) else return(z) end
  end
end

def sort(list,s,e)
#  target = list[s]
  target = list[s + (e-s)/2]
#  target = mid3(list[s],list[s + (e-s)/2],list[e])
  sp     = s
  ep     = e

  while true do
    while list[sp] < target && sp+1 < ep do
      sp += 1
    end
    
    while list[ep] >= target && sp+1 < ep do
      ep -= 1
    end

    if (list[sp] > list[ep])
      tmp      = list[ep]
      list[ep] = list[sp]
      list[sp] = tmp
    end

    if (sp + 1 == ep)
      break
    end
  end

  if ((sp - s) > 0)
    sort(list,s,sp)
  end
  if ((e - ep) > 0)
    sort(list,ep,e)
  end
end

#list = [ 10,9,8,7,6,5,4,3,2,1 ]
list = []
for no in 1..100000 do
  list.push(100000-no)
end

#pr(list)
sort(list,0,list.length-1)
#pr(list)

File: sort.pl
#! /usr/bin/perl

for($i=0;$i<100000;$i++) {
    push(@LIST,100000-$i);
}

#&pr();
&qsort(\@LIST,0,scalar(@LIST)-1);
#&pr();

sub mid3 {
  my($x,$y,$z) = @_;

  if ($x < $y) {
    if ($y < $z) {return($y);} elsif ($z < $x) {return($x);} else {return($z);}
  } 
  else {
    if ($z < $y) {return($y);} elsif ($x < $z) {return($x);} else {return($z);}
  }
}

sub qsort {
  my($LIST,$s,$e) =@_;

#  my($target) = $$LIST[$s];
  my($target) = $$LIST[$s + int(($e-$s)/2)];
#  my($target) = &mid3($$LIST[$s],$$LIST[int($s + ($e - $s)/2)],$$LIST[$e]);
  my($sp)     = $s;
  my($ep)     = $e;

  while(1) {
    while($$LIST[$sp]<$target && ($sp+1)<$ep) {
      $sp++;
    }
    while($$LIST[$ep]>=$target && ($sp+1)<$ep) {
      $ep--;
    }
    if ($$LIST[$sp]>$$LIST[$ep]) {
      my($tmp)    = $$LIST[$ep];
      $$LIST[$ep] = $$LIST[$sp];
      $$LIST[$sp] = $tmp;
    }
    if (($sp+1) == $ep) {
      last;
    }
  }

  if ($sp - $s > 0) {
    &qsort($LIST,$s,$sp);
  }

  if ($e - $ep > 0) {
    &qsort($LIST,$ep,$e);
  }
}

sub pr {
  printf("========\n");
  foreach $no (@LIST) {
    printf("%d\n",$no);
  }
}

